Some cards may return RSA signatures without leading zero bytes Fixes #1283

dengert opened the pull request:
Add leading zero bytes in a signature so it is the size of modulus.
Fixes #1283 

 On branch short-signatures
 Changes to be committed:
	modified:   libopensc/pkcs15-sec.c

<!--
Thank you for your pull request.


-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [ ] Tested with the following card: <!-- use `opensc-tool -n` to get the name of your card -->
	- [ ] tested PKCS#11
	- [ ] tested Windows Minidriver
	- [ ] tested macOS Tokend

mouse07410 commented:
Very preliminary test shows that this PR works with OpenPGP 2.0 (YubiKey NEO) and 2.1 (YubiKey 4).

dengert commented:
The fix for #1283 did not return the new length.  https://github.com/OpenSC/OpenSC/pull/1319/commits/0674dfef8f3035d9fa30a10039b24010f004e8c2 adds `r = modlen;`.  This needs to be tested with a card that fails, which I do not have.  

I can squash after https://github.com/MSoegtrop tests with failing card.

frankmorgner commented on src/libopensc/pkcs15-sec.c:
```diff
@@ -452,6 +452,16 @@ int sc_pkcs15_compute_signature(struct sc_pkcs15_card *p15card,
+	/* Some cards may return RSA signature as integer without leading zero bytes */
+	/* Already know outlen >= modlen and r >= 0 */
+	tmpoutlen = r;
+		if (obj->type == SC_PKCS15_TYPE_PRKEY_RSA && tmpoutlen < modlen) {
```
please fix the indenting and rewrite the code so that `tmpoutlen` is not needed.

frankmorgner commented:
What's the testing status, @MSoegtrop?

dengert commented:
Removed `temoutlen`  as requested and rebased into one commit. https://github.com/OpenSC/OpenSC/pull/1319/commits/dde41bbc063c056187a77c6b843ee34ec5902401
Tested with pkcs11-tool with card that does not have this problem. 

@MSoegtrop This still needs to be tested using a card that has this problem.

frankmorgner commented:
thanks
